WIP: Journeys card view

akwc10 opened the pull request.

akwc10 committed:
WIP Journeys Cardview

Added MarginItemDecoration to separate cards in recycler view.
Added CardviewFragment and corresponding ListAdapter for journeys.
Update AppTheme to MaterialComponents.
Added extra fields to domain model.
Added snackbar to allow retry when network error and made CoordinatorLayout root in case we add a FAB.

akwc10 committed:
Correction

fsbarata commented on app/src/main/java/com/my/journeyplanner/views/itineraryresults/ItineraryResultsCardviewListAdapter.kt:
```diff
@@ -0,0 +1,59 @@
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
+        MyViewHolder(
+            LayoutInflater.from(parent.context)
+                .inflate(R.layout.recycler_card_view_journey, parent, false)
```
you don't have to call this layout recycler_, it doesn't really care where you put it

akwc10 replied:
Renamed to card_view_journey

fsbarata commented on app/src/main/java/com/my/journeyplanner/views/itineraryresults/ItineraryResultsCardviewListAdapter.kt:
```diff
@@ -0,0 +1,59 @@
+||${Duration.between(leg.departureTime, leg.arrivalTime).toMinutes().toString()
+                .padStart(18, ' ')}${context.getString(R.string.mins)}
+""".trimMargin()
+        }
```
what's going on here lool. We probably should have a look at this to understand what you're trying to do

akwc10 replied:
Now uses linear layouts and dynamically adds the legs

fsbarata commented on app/src/main/java/com/my/journeyplanner/views/itineraryresults/ItineraryResultsCardviewListAdapter.kt:
```diff
@@ -0,0 +1,59 @@
+}
+
+class JourneyDiffCallback : DiffUtil.ItemCallback<Journey>() {
+    override fun areItemsTheSame(oldItem: Journey, newItem: Journey) = oldItem == newItem
```
is the identity of the Journey the whole object?

akwc10 replied:
Erm... not sure what you mean.

    data class Itinerary(
        val journeyVector: JourneyVector,
        val journeys: List<Journey>
    ) : JourneyPlannerResultDomainModel() {

        data class JourneyVector(
            val from: String,
            val to: String,
            val via: String
        ) : Serializable

        data class Journey(
            val arrivalDateTime: LocalDateTime,
            val legs: List<Leg>,
            val startDateTime: LocalDateTime
        ) : Serializable {
            data class Leg(
                val instructionSummary: String,
                val departureTime: LocalDateTime,
                val arrivalTime: LocalDateTime,
                val arrivalPoint: String,
                val departurePoint: String,
                val mode: String
            ) : Serializable
        }
    }

fsbarata replied:
we'll have to refine some of these things. This comparison you're doing here is extremely expensive.

akwc10 replied:
Now uses Journey.id() for diff callback override fun areItemsTheSame()

fsbarata commented on app/src/main/java/com/my/journeyplanner/views/itineraryresults/ItineraryResultsCardviewFragment.kt:
```diff
@@ -0,0 +1,82 @@
+const val DIVIDER_SPACE_HEIGHT = 32
+private val logger = KotlinLogging.logger {}
+
+class ItineraryResultsCardviewFragment : Fragment(), ItineraryResultsCardviewFragmentContract.View {
```
it doesn't really matter that it's a Cardview?

akwc10 replied:
Renamed to ItineraryResultsFragment after clearing out old class of the same name

fsbarata commented on app/src/main/java/com/my/journeyplanner/views/itineraryresults/ItineraryResultsCardviewFragment.kt:
```diff
@@ -0,0 +1,82 @@
+        super.onDetach()
+    }
+
+    interface OnItineraryResultsCardviewFragmentInteractionListener {
```
this name is too big mate. just name it Listener, it's already going to be called ItineraryResultsFragment.Listener because it's an inner interface, that's enough to be readable

akwc10 replied:
Renamed to Listener

akwc10 committed:
WIP Code review fixes

Various renaming improvements.
Removed old itinerary results journeys and legs.
Updating cardview from single text view to linear layouts.

akwc10 committed:
WIP Card view linear layout

akwc10 committed:
Added Journey.id() for diff callback override fun areItemsTheSame()

Cleaned up adding leg view.

akwc10 committed:
Correction

fsbarata commented on app/src/main/res/layout/card_view_journey.xml:
```diff
@@ -0,0 +1,45 @@
+                android:id="@+id/textViewStartAndArrivalTime"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:layout_weight="1" />
```
be careful how you use weights

akwc10 replied:
Removed unnecessary weight and corrected LinearLayout heights to wrap_content

akwc10 committed:
Corrected LinearLayout heights to wrap_content

Removed unnecessary weights.
